beacon-chain/powchain/service.go

```diff
@@ -364,20 +368,29 @@ func (w *Web3Service) ProcessDepositLog(depositLog gethTypes.Log) {
 // the ETH1.0 chain by trying to determine when to start the beacon chain.
 func (w *Web3Service) ProcessChainStartLog(depositLog gethTypes.Log) {
 	chainStartCount.Inc()
-	receiptRoot, timestampData, err := contracts.UnpackChainStartLogData(depositLog.Data)
+	chainStartDepositRoot, timestampData, err := contracts.UnpackChainStartLogData(depositLog.Data)
 	if err != nil {
 		log.Errorf("Unable to unpack ChainStart log data %v", err)
 		return
 	}
-	if w.depositTrie.Root() != receiptRoot {
-		log.Errorf("Receipt root from log doesn't match the root saved in memory,"+
-			" want %#x but got %#x", w.depositTrie.Root(), receiptRoot)
-		return
-	}
 
 	timestamp := binary.LittleEndian.Uint64(timestampData)
 	w.chainStarted = true
+	w.depositRoot = chainStartDepositRoot[:]
 	chainStartTime := time.Unix(int64(timestamp), 0)
+
+	// We then update the in-memory deposit trie from the chain start
+	// deposits at this point, as this trie will be later needed for
+	// incoming, post-chain start deposits.
+	sparseMerkleTrie, err := trieutil.GenerateTrieFromItems(
+		w.chainStartDeposits,
+		int(params.BeaconConfig().DepositContractTreeDepth),
+	)
+	if err != nil {
+		log.Fatalf("Unable to generate deposit trie from ChainStart deposits: %v", err)
+	}
+	w.depositTrie = sparseMerkleTrie
```

Comment:
We generate the Merkle trie from the chain start deposits after chain start is reached.